src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/CommandProcessor.java

```diff
@@ -482,6 +505,63 @@ public void doit(Tokens t) {
                 }
             }
         },
+        new Command("mem", "mem [-v] { address[/count] | address,address }", false) {
+            public void doit(Tokens t) {
+                int formatSize = (int)VM.getVM().getAddressSize();
+                boolean verbose = false;
+                String arg;
+
+                if (t.countTokens() == 2) {
+                    arg = t.nextToken();
+                    if (arg.equals("-v")) {
+                        verbose = true;
+                    } else {
+                        usage();
+                        return;
+                    }
+                }
+                if (t.countTokens() != 1) {
+                    usage();
+                    return;
+                }
+
+                arg = t.nextToken();
+                AddressRange addressRange = parseAddressRange(arg, formatSize);
+                if (addressRange == null) {
+                    usage();
+                    return;
+                }
+                Address start = addressRange.getStart();
+                Address end = addressRange.getEnd();
+
+                if (verbose) {
+                    // Do the equivalent of a findpc on the start address.
+                    PointerLocation loc = PointerFinder.find(start);
+                    loc.printOn(out);
+                }
+
+                int formatWidth = formatSize * 8 / 4 + 2;
+                while (start != null && start.lessThan(end)) {
+                    out.print(fillHexString(start, formatWidth));
+                    out.print(": ");
+                    Address val = start.getAddressAt(0);
+                    out.print(fillHexString(val, formatWidth));
+                    if (verbose) {
+                        // If we know what this is a pointer to, then print additional information.
+                        PointerLocation loc = PointerFinder.find(val);
+                        if (!loc.isUnknown()) {
+                            out.print(" ");
+                            loc.printOn(out, false, false);
+                        } else {
+                            out.println();
+                        }
+                    } else {
+                        out.println();
+                    }
```

Comment:
I wonder if it is possible to simplify the above fragment at 549-560 with something like:
                    if (verbose) {
                        // If we know what this is a pointer to, then print additional information.
                        PointerLocation loc = PointerFinder.find(val);
                        if (!loc.isUnknown()) {
                            out.print(" ");
                            loc.printOn(out, false, false);     // make it print without line end or ignore there is an extra end?
                        }
                    }
                    out.println();

Reply:
`printOn()` always prints a newline. It's hard to make it not do that since it relies on other `print` APIs, some of which print newlines and some of which don't (in which case `printOn()` will print a newline). So I can't easily just create a `printOn()` variant that doesn't print newlines, and printing two newlines would introduce a bunch of blank lines in the output, which wouldn't look good.

Reply:
Okay, thanks.